File: app/src/main/java/com/github/panda_io/micro_panda/ast/statement/switch.go
package statement

import (
	"github.com/panda-io/micro-panda/ast/ast"
	"github.com/panda-io/micro-panda/ast/ast_types"
	"github.com/panda-io/micro-panda/token"
)

type Switch struct {
	StatementBase
	Initialization ast.Statement
	Operand        ast.Expression
	Cases          []*Case
	Default        *Case
}

type Case struct {
	StatementBase
	Token token.Token
	Case  ast.Expression
	Body  ast.Statement
}

func (s *Switch) Validate(c ast.Context) {
	ctx := c.NewContext()
	if s.Initialization != nil {
		s.Initialization.Validate(ctx)
	}
	var operandType ast.Type
	if s.Operand == nil {
		c.Error(s.GetPosition(), "expect switch operand")
		return
	} else {
		s.Operand.Validate(ctx, nil)
		operandType = s.Operand.Type()
		if !ast_types.IsInteger(operandType) {
			c.Error(s.Operand.GetPosition(), "expect integer operand")
			return
		}
	}
	for _, ca := range s.Cases {
		caseCtx := ctx.NewContext()
		ca.Validate(caseCtx, operandType)
	}
	if s.Default != nil {
		defaultCtx := ctx.NewContext()
		s.Default.Validate(defaultCtx, operandType)
	}
}

func (c *Case) Validate(ctx ast.Context, operandType ast.Type) {
	if c.Case == nil {
		ctx.Error(c.GetPosition(), "expect case expression")
	} else {
		c.Case.Validate(ctx, operandType)
		if !c.Case.Type().Equal(operandType) {
			ctx.Error(c.GetPosition(), "case operand type mismatch with switch operand")
		}
	}
	if c.Body != nil {
		c.Body.Validate(ctx)
	}
}

File: app/src/main/java/com/github/panda_io/micro_panda/ast/statement/Switch.java
package com.github.panda_io.micro_panda.ast.statement;

import java.util.List;

import com.github.panda_io.micro_panda.ast.Context;
import com.github.panda_io.micro_panda.ast.Node;
import com.github.panda_io.micro_panda.ast.type.Type;
import com.github.panda_io.micro_panda.scanner.Token;

public class Switch extends Statement {
	public static class Case extends Node {
		public Token token;
		public com.github.panda_io.micro_panda.ast.expression.Expression caseExpr;
		public Statement body;

		public void validate(Context context, Type operandType) {
			if (this.caseExpr == null) {
				context.addError(this.getOffset(), "expect case expression");
			} else {
				this.caseExpr.validate(context, operandType);
				if (!this.caseExpr.getType().equal(operandType)) {
					context.addError(this.getOffset(), "case operand type mismatch with switch operand");
				}
			}
			if (this.body != null) {
				this.body.validate(context);
			}
		}
	}

	public Statement initialization;
	public com.github.panda_io.micro_panda.ast.expression.Expression operand;
	public List<Case> cases;
	public Case defaultCase;

	public void validate(Context context) {
		Context ctx = context.newContext();
		if (this.initialization != null) {
			this.initialization.validate(context);
		}
		Type operandType;
		if (this.operand == null) {
			context.addError(this.getOffset(), "expect switch operand");
			return;
		} else {
			this.operand.validate(context, null);
			operandType = this.operand.getType();
			if (!operandType.isInteger()) {
				context.addError(this.operand.getOffset(), "expect integer operand");
			}
		}
		for (Case caseNode : this.cases) {
			Context caseCtx = ctx.newContext();
			caseNode.validate(caseCtx, operandType);
		}
		if (this.defaultCase != null) {
			Context defaultCtx = ctx.newContext();
			this.defaultCase.validate(defaultCtx, operandType);
		}
	}
}
